### DISCUSSION TITLE
Keyword "validate" suggestion

### DISCUSSION THREAD

**Discussion by srburton** [dotnet/csharplang#1814] [Language Ideas]
_**I have suggestion process of new keywords for Sin data manipulation in C#**_ 

_**keywords:**_

`validate` : Keyword for validations of a variable .

`with` : keyword to signal which class will be initialized.

`any`: Accepts any given type as [int, string, bool ... ] Only when together with the keyword validate

**Example Class of a validator** 

```
public class MyValidate
{

  public void NotEmpty(validate any param,string errorMsg)
  {
       if(param == typeof(string))
       {
            if(param != String.Empty)
            {
                param += " very good";
            }
       }           

       throw new Exception(errorMsg); 

   }

  public void MaxCaracter(this string param,string errorMsg,int max)
  {
     //....
  }

}
```
**Example  validator**
```
string name = "c# is ";
validate name with (MyValidate) 
{
  NotEmpty("Not empty value!")
  MaxCaracter("Maximum caracter",8)
};
```
**or**

```
string name = "c# is ";
with name validate (MyValidate) 
{
  NotEmpty("Not empty value!")
  MaxCaracter("Maximum caracter",8)
};
```

**Output**
`Exception:  Maximum caracter`

In this example step a string to validate with less than 8 characters but in the validation is added more then when the second method is executed that counts the number of characters it already has a number greater than 8 then causes an exception

**Comment by Unknown6656** [comment]
What would be the advantage of your proposal compared to a combination of generics, pattern matching and maybe `dynamic` (yeah, yeah, I know it's slow and bad .....)?

You could rewrite `MyValidate.NotEmpty` as:
```csharp
public void NotEmpty<T>(T param, string errorMsg)
{
    if (param is string s && s != "")
    {
        s += " very good";

        // do something with s here (typesafe!)
    }

    throw new Exception(errorMsg);
}
```
or
```csharp
public void NotEmpty(dynamic param, string errorMsg)
{
    if (param is string s && s != "")
        param += " very good";

    // do something with param here (not typesafe)

    throw new Exception(errorMsg);
}
```

**Comment by srburton** [comment]
**KEYWORD  sequential** 

**Model People**
```
public class People
{
    public string Name {get; set;}
    public string LastName {get; set;}
    public string State {get; set;}
    public string City {get; set;}
}
```
**Validator methods**
```
public class Validator
{
     public static void NotEmpty<T>(T param, string error){  ...  }
     public static void MaxCaracter<T>(T param, string error,int max){  ...  }
     public static void CheckSocialSegure<T>(T socialNumber, T state, T city, string error) { ... }
}
```

**Object**

```

var people = new People();
var socialNumber = 232324;

sequential people in p 
{
   Validator.NotEmpty(p.Name,"Not Empty!");
   Validator.MaxCaracter(p,Name,"Max caracter is 8",8);
   Validator.NotEmpty(p.LastName,"Not Empty!");
   Validator.CheckSocialSegure(socialNumber, p.State, p.City, "Not is social segure!" );
}
```

**List**

```
var peoples = new List<People>()
{
    new People()
};
var socialNumber = 232324;

sequential peoples in p
{
   Validator.NotEmpty(p.Name,"Not Empty!");
   Validator.MaxCaracter(p,Name,"Max caracter is 8",8);
   Validator.NotEmpty(p.LastName,"Not Empty!");
   Validator.CheckSocialSegure(socialNumber, p.State, p.City, "Not is social segure!" );
}
```

**Primitive[]**
```
var myString = new string[] { "Hello", "C#"};

sequential myString in m
{
   Validator.NotEmpty(m,"Not Empty!");
}
```

**dynamic**
```
       var myDynamic = new
            {
                Name = "",
                LastName = "",
                Address = new
                {
                    City = ""
                }
            };
            

sequential myDynamic in m
{
   Validator.NotEmpty(m.Name,"Not Empty!");
   
   sequential m.Address in a
    {
             Validator.NotEmpty(a.City,"Not Empty city!");
             //Example "NameAndCity"
             //Validator.NameAndCity(m.Name,a.City,"City is valid!");
    }

}
```




In this example I want to demonstrate that a method if called within the _sequential_ it does not run on how much it does not reach its turn, this scenario would be for validations and testing in methods a native form of summarizing code and grouping a particular operation.

**Comment by AustinBryan** [comment]
I don't get the point of it at all. 

In your example:

```csharp
public class MyValidate
{
    public void NotEmpty(validate any param, string errorMsg)
    {
       if (param == typeof(string))
       {
            if (param != String.Empty)
                param += " very good";
       }           

       throw new Exception(errorMsg); 
    }
}
```

There's so many flaws here. For one, checking if `param` is a `string` and then you're adding `" very good"` if it is, and then you throw an exception, regardless if it is a `string` or not?

Also, for `param` to equal `typeof(string)` then it must be of type `Type` or a child class, which means `param += " very good"` probably wouldn't be valid. It depends, what `type` is `param`? Is it a `string`, `Type` or something that mimicks both interfaces? 

The biggest reason why I dislike the feature, is why not just do this:

```csharp
public class MyValidate
{
    public void NotEmpty(string param, string errorMsg)
    {
        if (param != String.Empty)
            param += " very good";
        throw new Exception(errorMsg); 
    }
}
```

Or use `object` or `dynamic` instead of making up so many keywords? I see literally nothing being gained here.

**Comment by srburton** [comment]
@AustinBryan   
My independent approach to code written above is that C# has something native to automation and data validation easily and simplistically I would like to contribute ideas so that we can find a good implementation for this functionality.

**Comment by CyrusNajmabadi** [comment]
Could you explain what you find insufficient in the existing mechanisms the language already has for "data validation"?

**Comment by srburton** [comment]
@CyrusNajmabadi 

**Current Code** 

```
public bool ValidEmailAddress(string emailAddress, out string errorMessage)
{
   // Confirm that the e-mail address string is not empty.
   if(emailAddress.Length == 0)
   {
      errorMessage = "e-mail address is required.";
         return false;
   }

   // Confirm that there is an "@" and a "." in the e-mail address, and in the correct order.
   if(emailAddress.IndexOf("@") > -1)
   {
      if(emailAddress.IndexOf(".", emailAddress.IndexOf("@") ) > emailAddress.IndexOf("@") )
      {
         errorMessage = "";
         return true;
      }
   }

   errorMessage = "e-mail address must be valid e-mail address format.\n" +
      "For example 'someone@example.com' ";
      return false;
}
```

**My approach** - *Implementation example*

In this case, validate returns a Boolean if all the test is true so ´ was satisfied with validation if a test is false then everything else is invalid!

```
public bool ValidEmailAddress(string emailAddress)
{
    
    return validate emailAddress in e
    {
          e.Lenght == 0 error "E-mail address is required."; 
          e.IndexOf("@") == -1 error "Email require @";
          e.Regex("pattern_valid_email") error "Email invalid!";
    }

}
```

**Comment by CyrusNajmabadi** [comment]
1. your example is very strange since you're checking different things in each form.  However, you could accomplish what you want in the second form with just:

```c#
          if (e.Lenght == 0) throw new ArgumentException("E-mail address is required.");
          if (e.IndexOf("@") == -1) throw new ArgumentException("Email require @");
          if (e.Regex("pattern_valid_email")) throw new ArgumentException("Email invalid!");
          return true;
```

This has the virtues of:

1. being idiomatic.
2. working in C# since the very first version of the language
3. being shorter than the form you specified.

**Comment by SilentSin** [comment]
@CyrusNajmabadi I think that example is supposed to return false if one of the tests fails (rather than throwing exceptions) and provide the error message as an out parameter. The problem being that the method doesn't appear to have an out parameter so I can't figure out of `error` is supposed to be that parameter (in which case it's just the current syntax without the if, parentheses, or equals sign) or if `error` is supposed to be a new keyword that somehow magics up an out parameter or something.

**Comment by CyrusNajmabadi** [comment]
in that case, you can just do:

```c#
          if (e.Lenght == 0) return (false, "E-mail address is required.");
          if (e.IndexOf("@") == -1) return (false, "Email require @");
          if (e.Regex("pattern_valid_email")) return (false, "Email invalid!");
          return (true, null);
```

**Comment by theunrepentantgeek** [comment]
It can also be done with a simple semantic type:

```
public ValidationResult ValidateEmailAddress(string emailAddress)
{
    return Validation.ErrorWhen( emailAddress.Length == 0, E-mail address is required.")
        && Validation.ErrorWhen( emailAddress.IndexOf("@") == -1, "Email require @")
        && Validation.ErrorWhen( emailAddress.Regex("pattern_valid_email"), "Email invalid!");
}
```
The `&&` does the usual short circuiting, so the expression exits early as soon as there is a problem.

This isn't theoretical code - it's written with a semantic type for validation that I wrote some time ago for my blog.

So, no need for special syntax in C# - just some library design. My `ValidationResult` is just one approach.
